Relocate variant index in WTF::CompactVariant

tuankiet65 opened the pull request:
#### df3e3ba21deda04897917f16f95844b5366c5a74
<pre>
Relocate variant index in WTF::CompactVariant
<a href="https://rdar.apple.com/149655111">rdar://149655111</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=291989">https://bugs.webkit.org/show_bug.cgi?id=291989</a>

Reviewed by Sam Weinig, David Degazio, and Darin Adler.

Relocate the variant index bits in WTF::CompactVariant from bit
63-56 to 55-48. Some architectures support ignoring the top bits
in a pointer (ex: AArch64 Top Byte Ignore), so this allows storing
the top byte untouched for applications that store metadata there.

* Source/WTF/wtf/CompactVariant.h:
* Source/WTF/wtf/CompactVariantOperations.h:
(WTF::CompactVariantOperations::decodedIndex):
(WTF::CompactVariantOperations::encodedPayload):
(WTF::CompactVariantOperations::encodedPayloadFromArguments):
* Tools/TestWebKitAPI/Tests/WTF/CompactVariant.cpp:
(TestWebKitAPI::TEST(WTF_CompactVariant, Pointers64)):
    - Added test to ensure the top byte in a 64-bit pointer is preserved.
      This test is only built on architectures with 64-bit pointers.

Canonical link: <a href="https://commits.webkit.org/294327@main">https://commits.webkit.org/294327@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c865aecdd6491a0fff3761e38657517bbb2d0d56

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101386 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21051 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11356 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106539 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52015 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103426 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21358 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29544 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77221 "Found 1 new test failure: media/video-src-blob-replay.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34259 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104393 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16481 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91576 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57563 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16302 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9588 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51363 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/94055 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86180 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9662 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108891 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/99996 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28515 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20980 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86198 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28877 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87778 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85757 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21829 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30484 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8195 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22632 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28446 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33738 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/123621 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28257 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34380 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31577 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29816 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddegazio reviewed (approved):
r=me

weinig reviewed (changes requested):
For the reasons stated above, marking this as needing more review.

weinig reviewed (approved):
Ok, nevermind, I see you changed the PR to no longer limit the number of alternatives.

darinadler commented on Source/WTF/wtf/CompactVariantOperations.h:
```diff
@@ -59,17 +59,24 @@ template<typename T> concept  CompactVariantAlternative =
+    static constexpr Storage indexShift = 48;
+    static constexpr Storage indexMask = 0xffUL << indexShift;
+    static constexpr Storage payloadMask = ~indexMask;
+    static constexpr Storage topByteMask = 0xff'00'00'00'00'00'00'00;
```
Pretty sure this won’t work correctly on 32-bit platforms. How do we expect it to know the right side is a 64-bit integer? In the past such constants have resulted in compiler errors on 32-bit platforms.

tuankiet65 replied:
I added `ULL` suffix so the literal is treated as `uint64_t`

darinadler commented on Tools/TestWebKitAPI/Tests/WTF/CompactVariant.cpp:
```diff
@@ -83,23 +83,39 @@ TEST(WTF_CompactVariant, Pointers)
 {
     int testInt = 1;
     float testFloat = 2.0f;
+    void* testPtr = reinterpret_cast<void*>(0x2600'1234'5678'9abc);
```
What does this do on 32-bit platforms?

tuankiet65 replied:
Yeah this test is only applicable on 64-bit platforms. I separated it out to a test gated with `CPU(ADDRESS64)` so it's only built on 64-bit platforms.

darinadler commented on Tools/TestWebKitAPI/Tests/WTF/CompactVariant.cpp:
```diff
@@ -83,23 +83,39 @@ TEST(WTF_CompactVariant, Pointers)
+    WTF::switchOn(variant,
+        [&](int* const& value)   { FAIL(); },
+        [&](float* const& value) { FAIL(); },
+        [&](void* const value) { EXPECT_EQ(reinterpret_cast<uintptr_t>(value), 0x2600'1234'5678'9abcUL); }
```
Pretty sure this will fail on 32-bit platforms.

tuankiet65 replied:
This test now only builds on 64-bit platforms.

darinadler commented on Source/WTF/wtf/CompactVariantOperations.h:
```diff
@@ -59,17 +59,24 @@ template<typename T> concept  CompactVariantAlternative =
-    static constexpr Storage payloadMask = (1ULL << payloadSize) - 1;
-    static_assert(payloadSize + indexSize <= totalSize);
+    static constexpr Storage indexShift = 48;
+    static constexpr Storage indexMask = 0xffULL << indexShift;
```
The `ULL` here isn't needed because the shift count is of type `Storage` and that causes promotion of the other operand.

tuankiet65 replied:
The compiler doesn't seem to like that, so I'll keep the `ULL`

```
CompactVariantOperations.h:77:30: error: constexpr variable 'indexMask' must be initialized by a constant expression
   77 |     static constexpr Storage indexMask = 0xFF << indexShift;
CompactVariantOperations.h:77:47: note: shift count 48 >= width of type 'int' (32 bits)
   77 |     static constexpr Storage indexMask = 0xFF << indexShift;
```

darinadler commented on Source/WTF/wtf/CompactVariantOperations.h:
```diff
@@ -59,17 +59,24 @@ template<typename T> concept  CompactVariantAlternative =
+    static constexpr Storage indexShift = 48;
+    static constexpr Storage indexMask = 0xffULL << indexShift;
+    static constexpr Storage payloadMask = ~indexMask;
+    static constexpr Storage topByteMask = 0xff'00'00'00'00'00'00'00ULL;
```
I'd prefer to sidestep the `ULL` and the assumption that it's 64 bits or more.
```suggestion
    static constexpr Storage topByteShift = 64 - 8;
    static constexpr Storage topByteMask = 0xff << topByteShift;
```
I also prefer uppercase hex `0xFF` but that's just a personal preference.

tuankiet65 replied:
I've changed to shifting as opposed to typing out the whole literal.

darinadler commented on Source/WTF/wtf/CompactVariantOperations.h:
```diff
@@ -78,7 +85,7 @@ template<CompactVariantAlternative... Ts> struct CompactVariantOperations {
     static constexpr Index decodedIndex(Storage value)
     {
-        return static_cast<Index>(static_cast<uint8_t>(value >> indexShift));
+        return static_cast<Index>(static_cast<uint8_t>((value & indexMask) >> indexShift));
```
Seems unnecessary to mask before shifting. I guess if we want to use `indexMask` or `payloadMask` here we could stick with doing it before shifting. Could shift first and mask with `0xFF`, which is `unshiftedIndexMask` (or we could call it `indexMask` and call the other one `shiftedIndexMask`).

tuankiet65 replied:
I don't want to introduce more constants so let's keep this.

darinadler commented on Source/WTF/wtf/CompactVariantOperations.h:
```diff
@@ -90,6 +97,19 @@ template<CompactVariantAlternative... Ts> struct CompactVariantOperations {
+        if constexpr (!CompactVariantAlternativePointer<T>) {
+            // Sanity check to make sure the data doesn't overwrite the top byte.
+            RELEASE_ASSERT(!(data & topByteMask));
+            data &= ~topByteMask;
```
This isn't needed. The `RELEASE_ASSERT` guarantees it's a no op.

tuankiet65 replied:
I've removed it.

darinadler commented on Source/WTF/wtf/CompactVariantOperations.h:
```diff
@@ -102,6 +122,19 @@ template<CompactVariantAlternative... Ts> struct CompactVariantOperations {
+        if constexpr (!CompactVariantAlternativePointer<T>) {
+            // Sanity check to make sure the data doesn't overwrite the top byte.
+            RELEASE_ASSERT(!(data & topByteMask));
+            data &= ~topByteMask;
```
This isn't needed. The `RELEASE_ASSERT` guarantees it's a no op.

tuankiet65 replied:
I've removed it.

darinadler commented on Source/WTF/wtf/CompactVariantOperations.h:
```diff
@@ -59,17 +59,24 @@ template<typename T> concept  CompactVariantAlternative =
-    static_assert(payloadSize + indexSize <= totalSize);
+    static constexpr Storage indexShift = 48;
+    static constexpr Storage indexMask = 0xffULL << indexShift;
+    static constexpr Storage payloadMask = ~indexMask;
```
I think it's overkill to have this as a separate named constant. I'd prefer we have just this or `indexMask` but not both.

tuankiet65 replied:
I like to have separate constants because it's more self-documenting.

darinadler reviewed (approved):
This seems OK as is but maybe a little more tweaking?

tuankiet65 committed:
Relocate variant index in WTF::CompactVariant
rdar://149655111
https://bugs.webkit.org/show_bug.cgi?id=291989

Reviewed by Sam Weinig, David Degazio, and Darin Adler.

Relocate the variant index bits in WTF::CompactVariant from bit
63-56 to 55-48. Some architectures support ignoring the top bits
in a pointer (ex: AArch64 Top Byte Ignore), so this allows storing
the top byte untouched for applications that store metadata there.

* Source/WTF/wtf/CompactVariant.h:
* Source/WTF/wtf/CompactVariantOperations.h:
(WTF::CompactVariantOperations::decodedIndex):
(WTF::CompactVariantOperations::encodedPayload):
(WTF::CompactVariantOperations::encodedPayloadFromArguments):
* Tools/TestWebKitAPI/Tests/WTF/CompactVariant.cpp:
(TestWebKitAPI::TEST(WTF_CompactVariant, Pointers64)):
    - Added test to ensure the top byte in a 64-bit pointer is preserved.
      This test is only built on architectures with 64-bit pointers.

Canonical link: https://commits.webkit.org/294327@main
